runtime/lua/vim/lsp/buf.lua

```diff
@@ -213,10 +213,15 @@ function M.format(options)
         return
       end
       local params = util.make_formatting_params(options.formatting_options)
-      client.request('textDocument/formatting', params, function(...)
+      client.request('textDocument/formatting', params, function(err, result, ctx, config)
         local handler = client.handlers['textDocument/formatting']
           or vim.lsp.handlers['textDocument/formatting']
-        handler(...)
+        local mode = vim.api.nvim_get_mode().mode
+        if changedtick ~= vim.api.nvim_buf_get_var(ctx.bufnr, "changedtick") or mode:find('^i') then
```

Comment:
What if one wants to use `vim.lsp.buf.format()` from Insert mode or `i_CTRL-O`?

Reply:
In that case `async` doesn't make sense, I think. Your insert mode changes would be overwritten. You'd have to not do anything until formatting is done, which defeats the purpose of async.

Reply:
> Your insert mode changes would be overwritten. 

How so? Isn't changedtick also updated in insert mode?

> You'd have to not do anything until formatting is done.

You'd have to not **change** anything. That's the case regardless of mode, no?

I fail to see why insert mode should be a special case. What if the user wants to trigger format then go to normal mode or just move the cursor while in insert mode? You should still be able to do those things without that invalidating the request (because it's still valid, and the buffer has not changed).